### PR TITLE
Ajay tripathy add gpu support

### PR DESCRIPTION
**Pull request by AjayTripathy**
## What does this PR change?
* Adds Custom Pricing for GPU support

## Does this PR relate to any other PRs?
* No

## How will this PR impact users?
* Allow nvidia.com/GPU support for on-prem deployments

## How was this PR tested?
* Unit tests

## Does this PR require changes to documentation?
* We should document adding CSV support for on-prem GPU deployments

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next Opencost release? If not, why not?
* Yes

### CODE REVIEW

**Comment by dwbrown2** [comment]
@AjayTripathy can we make sure to update docs as part of this?

@jessegoodier @srpomeroy as FYI

**Comment by michaelmdresser** [comment]
This Git history is pretty crazy, did you perhaps base this branch off of `v1.98` rather than `develop`?

File: pkg/cloud/csvprovider.go [hunk 1]
```diff
@@ -299,9 +338,16 @@ func PVValueFromMapField(m string, n *v1.PersistentVolume) string {
 
 func (c *CSVProvider) GetKey(l map[string]string, n *v1.Node) Key {
 	id := NodeValueFromMapField(c.NodeMapField, n, c.UsesRegion)
+	var gpuCount int64
+	gpuCount = 0
+	if gpuc, ok := n.Status.Capacity["nvidia.com/gpu"]; ok { // TODO: support non-nvidia GPUs
```
Comment (jessegoodier) [inline pkg/cloud/csvprovider.go:346]: maybe we just something like this:
kubecost.com/gpu-type

good as-is for my needs though. Thanks Ajay!

Comment (srpomeroy) [reply pkg/cloud/csvprovider.go:346]: @jessegoodier Did you end up testing in the cloud with a GPU instance to see if there is a standard tag/label?

Comment (srpomeroy) [reply pkg/cloud/csvprovider.go:343]: nm. Looks like this is the suggested label from Kubernetes docs.
https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/

Comment (jessegoodier) [reply pkg/cloud/csvprovider.go:346]: feedback from another on-prem customer:
`gpu.nvidia.com/class`

there are no annotations in AWS for the node.

essentially no standard out there.

This customer has thousands of nodes with the above label though. Can we support both?

Comment (jessegoodier) [reply pkg/cloud/csvprovider.go:346]: so the two node annotations we have requests for:
```
nvidia.com/gpu_type=Quadro_RTX_4000
gpu.nvidia.com/gpu_type=Quadro_RTX_4000
```

Comment (AjayTripathy) [reply pkg/cloud/csvprovider.go:346]: Thanks for the feedback. This PR doesn't support multiple yet but it can be done.

**Comment by jessegoodier** [comment]
question: how will multiple GPUs be handled?

As of now, all customers are running homogenous models per node (all 1 gpu-type)

A container can use multiple as you can see below, and kubecost is already aware of this.

default pricing is $693.576 in `defaultModelPricing`

Based on the output below, GPU count is ignored?

 rate $693.576 x 41.3333 / 720 = $39.81637 (should be x4)

Output from a customer:
```
        "start": "2022-11-02T20:14:00Z",
        "end": "2022-11-03T06:34:00Z",
        "minutes": 620,
        "cpuCores": 40,
        "cpuCoreRequestAverage": 40,
        "cpuCoreUsageAverage": 9.665386,
        "cpuCoreHours": 413.333333,
        "cpuCost": 15.89928,
        "cpuCostAdjustment": 0,
        "cpuEfficiency": 0.241635,
        "gpuCount": 4,
        "gpuHours": 41.333333,
        "gpuCost": 39.270965,
```

File: pkg/cloud/csvprovider.go (hunk 1 above)
Comment (jessegoodier) [reply pkg/cloud/csvprovider.go:346]: Should this line be ["nvidia.com/gpu_type"]?

image

Comment (AjayTripathy) [reply pkg/cloud/csvprovider.go:346]: No, this is how capacity is counted, not the type of the gpu.

**Comment by AjayTripathy** [comment]
> question: how will multiple GPUs be handled?
> 
> As of now, all customers are running homogenous models per node (all 1 gpu-type)
> 
> A container can use multiple as you can see below, and kubecost is already aware of this.
> 
> default pricing is $693.576 in `defaultModelPricing`
> 
> Based on the output below, GPU count is ignored?
> 
> rate $693.576 x 41.3333 / 720 = $39.81637 (should be x4)
> 
> Output from a customer:
> 
> ```
>         "start": "2022-11-02T20:14:00Z",
>         "end": "2022-11-03T06:34:00Z",
>         "minutes": 620,
>         "cpuCores": 40,
>         "cpuCoreRequestAverage": 40,
>         "cpuCoreUsageAverage": 9.665386,
>         "cpuCoreHours": 413.333333,
>         "cpuCost": 15.89928,
>         "cpuCostAdjustment": 0,
>         "cpuEfficiency": 0.241635,
>         "gpuCount": 4,
>         "gpuHours": 41.333333,
>         "gpuCost": 39.270965,
> ```

This should be updated (and fixed) in this PR!

**Comment by AjayTripathy** [comment]
CLosing in favor of https://github.com/opencost/opencost/pull/1480 with correct git history

**Pull request closed without merging**